numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -2923,8 +2923,36 @@ static PyObject *
 
     /* TODO: include type name in error message, which is not @name@ */
     PyObject *obj = NULL;
+#if defined(_@TYPE@_IS_CDOUBLE) || defined(_@TYPE@_IS_CFLOAT)
+    static char *kwnames[] = {"", "", NULL};  /* positional-only */
+    static char *parse_error = "Could not convert arguments into a complex scalar. '%R' given.";
+    PyObject *real_obj = NULL;
+    PyObject *imag_obj = NULL;
+    if (PyArg_ParseTupleAndKeywords(args, kwds, "|OO", kwnames, &real_obj, &imag_obj)) {
+        if (imag_obj == NULL) {
+            obj = real_obj;
+            Py_XINCREF(obj);
+        }
+        else if (PyNumber_Check(real_obj) && PyNumber_Check(imag_obj)) {
+            if (PyComplex_Check(real_obj) || PyComplex_Check(imag_obj)) {
+                PyErr_Format(PyExc_TypeError, parse_error, args);
+                return NULL;
+            }
+            obj = PyComplex_Type.tp_new(&PyComplex_Type, args, NULL);
+            if (obj == NULL) {
+                return NULL;
+            }
+        }
+        else {
+            PyErr_Format(PyExc_TypeError, parse_error, args);
+            return NULL;
```

Comment:
My slight preference would be to put the error paths first, then no indent for the successful one.

Reply:
Refactored so the error case comes first.